staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/convertlastapplied.go

```diff
@@ -0,0 +1,137 @@
+// so that upgrading from client-side apply to server-side apply doesn't conflict
+// by default because the user intends to manage the object with server-side apply now
+func (f *convertLastAppliedManager) Apply(liveObj runtime.Object, newObj runtime.Object, managed Managed, manager string, force bool) (runtime.Object, Managed, error) {
+	// Upgrade the client-side apply annotation
+	// only if the manager is the default manager for kubectl server-side apply
+	if manager != "kubectl" {
+		return f.fieldManager.Apply(liveObj, newObj, managed, manager, force)
+	}
+
+	lastAppliedFields, err := f.convertLastAppliedToFieldSet(liveObj)
+	// TODO: if there's an error, what do we do? should we remove the CSA annotation?
+	if err != nil {
+		return f.fieldManager.Apply(liveObj, newObj, managed, manager, force)
+	}
+	fields := lastAppliedFields
+	existing, ok := managed.Fields()[manager]
+	if ok {
+		fields = existing.Set().Union(lastAppliedFields)
+	}
+	var apiVersion = fieldpath.APIVersion(f.groupVersion.String())
+	managed.Fields()[manager] = fieldpath.NewVersionedSet(fields, apiVersion, false)
+
+	return f.fieldManager.Apply(liveObj, newObj, managed, manager, force)
+}
+
+func (f *convertLastAppliedManager) convertLastAppliedToFieldSet(liveObj runtime.Object) (*fieldpath.Set, error) {
+	var accessor, err = meta.Accessor(liveObj)
+	if err != nil {
+		panic(fmt.Sprintf("couldn't get accessor: %v", err))
+	}
+
+	// If there is no client-side apply annotation, then there is nothing to do
+	var annotations = accessor.GetAnnotations()
+	if annotations == nil {
+		return nil, nil
+	}
+	var lastApplied, ok = annotations[corev1.LastAppliedConfigAnnotation]
+	if !ok || lastApplied == "" {
+		return nil, nil
+	}
+
+	liveObjTyped, err := f.typeConverter.ObjectToTyped(liveObj)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert live obj to typed: %v", err)
+	}
+
+	var lastAppliedObj = &unstructured.Unstructured{Object: map[string]interface{}{}}
+	err = json.Unmarshal([]byte(lastApplied), lastAppliedObj)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode last applied obj: %v", err)
+	}
+
+	if lastAppliedObj.GetAPIVersion() != f.groupVersion.String() {
+		return nil, fmt.Errorf("expected version of last applied to match live object '%s', but got '%s': %v", f.groupVersion.String(), lastAppliedObj.GetAPIVersion(), err)
+	}
+
+	lastAppliedObjTyped, err := f.typeConverter.ObjectToTyped(lastAppliedObj)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert last applied to typed: %v", err)
+	}
+
+	lastAppliedObjFieldSet, err := lastAppliedObjTyped.ToFieldSet()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create fieldset for last applied object: %v", err)
+	}
+
+	comparison, err := lastAppliedObjTyped.Compare(liveObjTyped)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compare last applied object and live object: %v", err)
+	}
+
+	lastAppliedObjFieldSet = lastAppliedObjFieldSet.
```

Comment:
I was a little mislead by the function name, since I was expecting to strictly get the `lastAppliedFieldSet`. This is not exactly the same thing.

Reply:
Renamed to `allowedConflictsFromLastApplied`